types/+MPCT/code_MPCT_ADMM_semiband_C.c

```diff
@@ -0,0 +1,930 @@
+    }
+    for(unsigned int i = 0; i < mm_; i++){
+        ur[i] = ur_in[i];
+    }
+    #endif
+
+    // Compute b
+
+    for(unsigned int i = 0 ; i< nn_ ; i++){
+        b[i] = x0[i];
+    }
+
+
+    // Compute q
+
+    for(unsigned int i = 0 ; i < nn_ ; i++){
+        
+        for(unsigned int j = 0 ; j < nn_ ; j++){
+
+            q[NN_*nm_ + i] -= T[i][j] * xr[j];
+
+        }
+
+    }
+
+    for(unsigned int i = 0 ; i < mm_ ; i++){
+        
+        for(unsigned int j = 0 ; j < mm_ ; j++){
+
+            q[NN_*nm_ + nn_ + i] -= S[i][j] * ur[j];
+
+        }
+
+    }
+
+    // Measure time
+    #if MEASURE_TIME == 1
+
+    #if WIN32
+    QueryPerformanceCounter(&post_update); // Get time after the update    
+    t_update_time = 1000000000ULL * (post_update.QuadPart - start.QuadPart) / frequency.QuadPart;
+    sol->update_time = t_update_time/(double)1e+9;
+    #else // If Linux
+    clock_gettime(CLOCK_MONOTONIC_RAW, &post_update);
+    sol->update_time = (double) ( (post_update.tv_sec - start.tv_sec) * 1000.0 ) + (double) ( (post_update.tv_nsec - start.tv_nsec) / 1000000.0 );
+    #endif
+
+    #endif
+
+    // Algorithm
+    while(done == 0){
+
+        k += 1;
+
+        // Save the value of v
+        memcpy(v_old, v, sizeof(double)*(NN_+1)*nm_);
+        
+        // Reset acumulator variables
+        memset(z1_ac, 0, sizeof(double)*(NN_+1)*nm_);
+        memset(z3_ac, 0, sizeof(double)*(NN_+1)*nm_);
+        memset(z1_b, 0, sizeof(double)*(NN_+2)*nn_);
+        memset(z3_b, 0, sizeof(double)*(NN_+2)*nn_);
+        memset(z2, 0, sizeof(double)*2*nm_); 
+
+        //********** Equality-constrained QP solve **********//
+        // This problem updates z
+        for(unsigned int i = 0; i < (NN_+1)*nm_ ; i++){
+
+            #ifdef SCALAR_RHO
+            p[i] = q[i] + lambda[i] - rho * v[i];
+            #else
+            p[i] = q[i] + lambda[i] - rho[i] * v[i];
+            #endif
+
+        }
+
+
+        /****** Compute xi from eq. (9a) using Alg. 2 from the article ******/
+
+        solve_banded_diag_sys(Q_rho_i, R_rho_i, S_rho_i, T_rho_i, z1_ac, p); // Obtains z1_a
+
+        // z2_a = M_hat * z1_a computed sparsely
+        for (unsigned int i = 0 ; i < nn_ ; i++){
+
+            for (unsigned int l = 0 ; l < NN_+1 ; l++){
+
+                for (unsigned int j = 0 ; j < nn_ ; j++){
+
+                    if (l == NN_){
+                        z2[i] += M_hat_x[i][j+nn_]*z1_ac[l*nm_+j];
+                    }
+                    else{
+                        z2[i] += M_hat_x[i][j]*z1_ac[l*nm_+j];
+                    }
+                    
+                }
+                
+            }
+
+        }
+
+        for (unsigned int i = nn_ ; i < nm_ ; i++){
+
+            for (unsigned int l = 0 ; l < NN_+1 ; l++){
+            
+                for (unsigned int j = 0 ; j < mm_ ; j++){
+                    if (l == NN_){
+                        z2[i] += M_hat_u[i-nn_][j+mm_]*z1_ac[l*nm_+nn_+j];
```

Comment:
The additions and subtractions when accessing the `M_hat_x` and `M_hat_u` matrices might be avoidable if more variables are declared. For example, the objective would be to have `M_hat_u_2[i][j]` than the current `M_hat_u[i-nn_][j+mm_]` that you can see in this line. This might also help the compiler.

Reply:
I have done some modifications on this. Maybe it could be improved even more.